### PR TITLE
Retire Project components

### PR DESCRIPTION
**Pull request by i-walker**
Project components as-is are deprecated in IntelliJ. 
Depending on their usage they fall in different migrating categories. 

The behavior of the CacheLoader suggests a Listener as it does nothing up to the life cycle phase,  where the Ide spins up a Haskell project. 
The listener also enjoys the advantage to be disposed automatically.

### CODE REVIEW

**Comment by i-walker** [comment]
From the current state I am getting 3 failed tests, related to the parser:  
Screenshot 2020-05-13 at 19 13 00

File: src/com/haskforce/cabal/highlighting/CabalAnnotator.scala [hunk 1]
```diff
@@ -20,8 +20,7 @@ class CabalAnnotator extends Annotator {
 
   private def setHighlighting(element: PsiElement, holder: AnnotationHolder, key: TextAttributesKey) {
     holder
-      .newSilentAnnotation(HighlightSeverity.INFORMATION)
-      .enforcedTextAttributes(EditorColorsManager.getInstance.getGlobalScheme.getAttributes(key))
-      .create()
+        .createWeakWarningAnnotation(element, "Message")
+      .setEnforcedTextAttributes(EditorColorsManager.getInstance.getGlobalScheme.getAttributes(key))
```
Comment (carymrobbins) [inline src/com/haskforce/cabal/highlighting/CabalAnnotator.scala:24]: Is this change intended to be here?

File: src/com/haskforce/codeInsight/HaskellCompletionCacheLoader.scala [hunk 2]
```diff
@@ -7,38 +7,26 @@ import com.haskforce.psi.{HaskellFile, HaskellPsiUtil}
...
+import com.intellij.openapi.project.{Project, ProjectManagerListener}
 import com.intellij.openapi.util.Computable
 import com.intellij.openapi.vfs._
 import com.intellij.psi.{PsiFile, PsiManager}
 
 import scala.annotation.tailrec
 
 /** Loads the completion cache for any Haskell files we have open. */
-class HaskellCompletionCacheLoader(project: Project) extends ProjectComponent {
+class HaskellCompletionCacheLoader extends ProjectManagerListener {
```
Comment (carymrobbins) [inline src/com/haskforce/codeInsight/HaskellCompletionCacheLoader.scala:20]: I think this might require changes to the plugin.xml?

File: src/com/haskforce/internal/HaskForceInternal.scala [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    val app = ApplicationManager.getApplication()
+    registerTopic(app, ProjectManager.TOPIC, new HaskellCompletionCacheLoader)
+  }
+
+  def registerTopic[A](app: Application, topic: Topic[A], instance: A): Unit = {
+    app.getMessageBus.connect(app).subscribe(topic, instance)
+  }
+}
+
+
```
Comment (carymrobbins) [inline src/com/haskforce/internal/HaskForceInternal.scala:20]: Seems like this might need to be registered in the plugin.xml as well. Also, I think this might be better named as `com.haskforce.ide.HaskForceInitializedListener` to match the interface. Plus, `internal` seems a little vague for what this class is used for.

File: src/com/haskforce/codeInsight/HaskellCompletionCacheLoader.scala (hunk 2 above)
Comment (i-walker) [reply src/com/haskforce/codeInsight/HaskellCompletionCacheLoader.scala:20]: Yes, I forgot to register the ApplicationInitialized instance. I'll add that

File: src/com/haskforce/internal/HaskForceInternal.scala (hunk 3 above)
Comment (i-walker) [reply src/com/haskforce/internal/HaskForceInternal.scala:20]: we can rename it to whatever makes sense to you

File: src/com/haskforce/cabal/highlighting/CabalAnnotator.scala (hunk 1 above)
Comment (i-walker) [reply src/com/haskforce/cabal/highlighting/CabalAnnotator.scala:24]: yes as it  does not compile otherwise

Comment (carymrobbins) [reply src/com/haskforce/cabal/highlighting/CabalAnnotator.scala:24]: Strange, those changes fail to compile for me due to their deprecation -

```
:compileScala
.../intellij-haskforce/src/com/haskforce/cabal/highlighting/CabalAnnotator.scala:23: method createWeakWarningAnnotation in trait AnnotationHolder is deprecated: see corresponding Javadoc for more information.
        .createWeakWarningAnnotation(element, "Message")
         ^
one error found
```

File: src/com/haskforce/codeInsight/HaskellCompletionCacheLoader.scala (hunk 2 above)
Comment (carymrobbins) [inline src/com/haskforce/codeInsight/HaskellCompletionCacheLoader.scala:20]: There's also another issue. In the companion object there's this

```scala
object HaskellCompletionCacheLoader {
  def get(project: Project): HaskellCompletionCacheLoader = {
    project.getComponent(classOf[HaskellCompletionCacheLoader])
  }
```

That obviously won't work anymore. Need to find a better way to do this. The `cache` field needs to be scoped by project, which isn't quite addressed in this PR yet.

File: resources/META-INF/plugin.xml [hunk 4]
```diff
@@ -440,6 +432,8 @@
   </actions>
 
   <extensions defaultExtensionNs="com.intellij">
+     <!-- ApplicationInitializedListener  -->
```
Comment (carymrobbins) [inline resources/META-INF/plugin.xml:435]: I think the comment is redundant since it's identical to the extension point name.

File: src/com/haskforce/codeInsight/HaskellCompletionCacheLoader.scala (hunk 2 above)
Comment (i-walker) [reply src/com/haskforce/codeInsight/HaskellCompletionCacheLoader.scala:20]: no problem. I'll surface it as a service. Then it is scoped and distributed to all projects

Comment (i-walker) [reply src/com/haskforce/codeInsight/HaskellCompletionCacheLoader.scala:20]: I see this part was bundling state with the component. That has to be refactored into a separate service, that will manage and store the cache. 
I am only voicing this so you understand the steps I am taking

Comment (i-walker) [reply src/com/haskforce/codeInsight/HaskellCompletionCacheLoader.scala:20]: In the future versions of Intellij that has to be separated. 
pure Morphisms can be distributed as services with functions that can be overridden by 3rd parties depending on this plugin. If you choose to. 
IntelliJ also allows them to contain mutable variables or state, but personally I'd like to keep them apart, as this allows one to reason over simpler.

File: src/com/haskforce/cabal/highlighting/CabalAnnotator.scala (hunk 1 above)
Comment (i-walker) [reply src/com/haskforce/cabal/highlighting/CabalAnnotator.scala]: should be removed now
